src/main/java/org/springframework/integration/aws/outbound/SnsMessageHandler.java

```diff
@@ -0,0 +1,197 @@
+		this.topicArnExpression = topicArnExpression;
+	}
+
+	public void setSubject(String subject) {
+		Assert.hasText(subject, "subject must not be empty.");
+		this.subjectExpression = new LiteralExpression(subject);
+	}
+
+	public void setSubjectExpression(Expression subjectExpression) {
+		Assert.notNull(subjectExpression, "subjectExpression must not be null.");
+		this.subjectExpression = subjectExpression;
+	}
+
+	/**
+	 * The {@link Expression} to produce the SNS notification message.
+	 * If it evaluates to the {@link SnsBodyBuilder} the {@code messageStructure}
+	 * of the {@link PublishRequest} is set to {@code json}.
+	 * Otherwise the {@link #getConversionService()} is used to convert the evaluation result
+	 * to the {@link String} without setting the {@code messageStructure}.
+	 * @param bodyExpression the {@link Expression} to produce the SNS notification message.
+	 */
+	public void setBodyExpression(Expression bodyExpression) {
+		Assert.notNull(bodyExpression, "bodyExpression must not be null.");
+		this.bodyExpression = bodyExpression;
+	}
+
+	@Override
+	protected void onInit() throws Exception {
+		super.onInit();
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+		TypeLocator typeLocator = this.evaluationContext.getTypeLocator();
+		if (typeLocator instanceof StandardTypeLocator) {
+				/*
+				 * Register the 'org.springframework.integration.aws.support' package
+				 * you don't need a FQCN for the 'SnsMessageBuilder'.
+				 */
+			((StandardTypeLocator) typeLocator).registerImport("org.springframework.integration.aws.support");
+		}
+	}
+
+	@Override
+	protected void handleMessageInternal(Message<?> message) throws Exception {
+		Object payload = message.getPayload();
+
+		PublishRequest publishRequest = null;
+
+		if (payload instanceof PublishRequest) {
+			publishRequest = (PublishRequest) payload;
+		}
+		else {
+			publishRequest = new PublishRequest();
+			if (this.topicArnExpression != null) {
+				String topicArn = this.topicArnExpression.getValue(this.evaluationContext, message, String.class);
+				publishRequest.setTopicArn(topicArn);
+			}
+
+			if (this.subjectExpression != null) {
+				String subject = this.subjectExpression.getValue(this.evaluationContext, message, String.class);
+				publishRequest.setSubject(subject);
+			}
+
+			Object snsMessage = message.getPayload();
+
+			if (this.bodyExpression != null) {
+				snsMessage = this.bodyExpression.getValue(this.evaluationContext, message);
+			}
+
+			if (snsMessage instanceof SnsBodyBuilder) {
+				publishRequest.withMessageStructure("json")
+						.setMessage(((SnsBodyBuilder) snsMessage).build());
+			}
+			else {
+				publishRequest.setMessage(getConversionService().convert(snsMessage, String.class));
+			}
+		}
+
+		PublishResult publishResult = this.amazonSns.publish(publishRequest);
+
+		if (this.applicationEventPublisher != null) {
```

Comment:
Because it has only one `messageId` field. So, the whole `PublishResult` is redundant and it doesn't have any correlation with that `PublishRequest`. According to the hard logic above for the `PublishRequest`, I find it as overhear to `reply` just with the `messageId` and make end-user to re-build `PublishResult` against `requestMessage` one more time.